src/Microsoft.OData.Core/Uri/ApplyClauseToStringBuilder.cs

```diff
@@ -0,0 +1,170 @@
+                _query.Append(ExpressionConstants.SymbolEscapedSpace);
+                AppendWord(ExpressionConstants.KeywordWith);
+                AppendWord(GetAggregationMethodName(aggExpression));
+                AppendWord(ExpressionConstants.KeywordAs);
+                _query.Append(aggExpression.Alias);
+            }
+        }
+
+        private void Translate(FilterTransformationNode transformation)
+        {
+            AppendExpression(transformation.FilterClause.Expression);
+        }
+
+        private void Translate(GroupByTransformationNode transformation)
+        {
+            bool appendComma = false;
+            foreach (GroupByPropertyNode node in transformation.GroupingProperties)
+            {
+                if (appendComma)
+                    AppendComma(appendComma);
+                else
+                {
+                    appendComma = true;
+                    _query.Append(ExpressionConstants.SymbolOpenParen);
+                }
+
+                if (node.Expression != null)
+                    AppendExpression(node.Expression);
+
+                bool appendSlash = false;
+                foreach (GroupByPropertyNode childNode in node.ChildTransformations)
+                {
+                    appendSlash = AppendSlash(appendSlash);
+                    AppendExpression(childNode.Expression);
+                }
+            }
+            if (appendComma)
+                _query.Append(ExpressionConstants.SymbolClosedParen);
+
+            if (transformation.ChildTransformations != null)
+            {
+                AppendComma(true);
+                Translate(transformation.ChildTransformations);
+            }
+        }
+
+        public string TranslateApplyClause(ApplyClause applyClause)
+        {
+            ExceptionUtils.CheckArgumentNotNull(applyClause, nameof(applyClause));
+
+            _query.Append(ExpressionConstants.QueryOptionApply);
+            _query.Append(ExpressionConstants.SymbolEqual);
+
+            bool appendSlash = false;
+            foreach (TransformationNode transformation in applyClause.Transformations)
+            {
+                appendSlash = AppendSlash(appendSlash);
+                Translate(transformation);
+            }
+
+            return appendSlash ? _query.ToString() : string.Empty;
+        }
+
+        private void Translate(TransformationNode transformation)
+        {
+            switch (transformation.Kind)
+            {
+                case TransformationNodeKind.Aggregate:
+                    _query.Append(ExpressionConstants.KeywordAggregate);
+                    break;
+                case TransformationNodeKind.GroupBy:
+                    _query.Append(ExpressionConstants.KeywordGroupBy);
+                    break;
+                case TransformationNodeKind.Filter:
+                    _query.Append(ExpressionConstants.KeywordFilter);
+                    break;

```

Comment:
Let's add $compute processing here (and a new corresponding value in enum TransformationNodeKind), so that we can close the round-trip gap for the introduction of $compute query option, which is also supported inside $apply (e.g. Get ~/Sales?$apply=computer(Amount mul Product/TaxRate as Tax, see [http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs02/odata-data-aggregation-ext-v4.0-cs02.html#_Toc435016588](http://docs.oasis-open.org/odata/odata-data-aggregation-ext/v4.0/cs02/odata-data-aggregation-ext-v4.0-cs02.html#_Toc435016588)).

I understand that we don't need to implement other transformations called by this OData Aggregation Extension except for these four transformations (aggregate, groupby, filter and compute).

It will be great if you can take care of this in this PR related to "apply".
Thanks!

Reply:
implement $apply compute in https://github.com/OData/odata.net/pull/1050/commits/bf18e56807babf55e360233f526cfdf7a09d16e9